### PR TITLE
resource_manager: make limit effective quickly and avoid jitter

### LINKED ISSUE #5946: Quota Limit have a jitter after change the resource group config
### What did you do?
300 ru_per_sec ->  900 ru_per_sec
image

### What did you expect to see?

qps increase from 1800 -> 4000

### What did you see instead?
qps 1800 -> 7400 -> 4000

### LINKED ISSUE #5954: restart TiDB or idle for a while, resource control cannot limit quickly
## Development Task
image
here lasted 40s

### LINKED ISSUE #5955: Too many resource group creation log
## Development Task
FhTogxu5II

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: close #5955 close #5954 close #5946

### What is changed and how does it work?
Client side
1.  When controller want to create groupCostController, the goroutine should acquire the lock firstly.
2. Reduce tokenCounter's initial speed to 0 to avoid requesting too many tokens
3. Set the NotifyThreshold to a value close to the initialRequestUnits so that request tokens as soon as possible when token consumption occurs.

Server side
1. Add a setttingChanged parameter to detect whether the token bucket configuration has changed. If the token bucket configuration has changed and there is a loan, the loan will be reset
2. Set the LoanCoefficient to 1 when FillRate is greater than BurstLimit

image
```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by nolouch — COMMENTED** [review]
plz make the title and commit message be meaningful :(

File: client/resource_manager/client/client.go [hunk 1]
```diff
@@ -133,15 +136,25 @@ func (c *ResourceGroupsController) Stop() error {
 }
 
 func (c *ResourceGroupsController) putResourceGroup(ctx context.Context, name string) (*groupCostController, error) {
+	// ref https://github.com/tikv/pd/issues/5955
+	c.createGroupsControllerLock.Lock()
+	defer c.createGroupsControllerLock.Unlock()
```
Comment (JmPotato) [inline client/resource_manager/client/client.go:141]: Maybe there is no need to introduce a lock here, we can just use `LoadOrStore` when we try to store the group into `c.groupsController`? If it already exists, just do nothing.

File: client/resource_manager/client/client.go [hunk 2]
```diff
@@ -153,7 +169,10 @@ func (c *ResourceGroupsController) updateAllResourceGroups(ctx context.Context)
 	latestGroups := make(map[string]struct{})
 	for _, group := range groups {
 		log.Info("create resource group cost controller", zap.String("name", group.GetName()))
-		gc := newGroupCostController(group, c.config, c.lowTokenNotifyChan, c.groupNotificationCh)
+		gc, err := newGroupCostController(group, c.config, c.lowTokenNotifyChan, c.groupNotificationCh)
+		if err != nil {
+			continue
```
Comment (JmPotato) [inline client/resource_manager/client/client.go:174]: Do we need to print a log here?

File: client/resource_manager/client/limiter.go [hunk 3]
```diff
@@ -62,9 +62,9 @@ func Every(interval time.Duration) Limit {
 // or its associated context.Context is canceled.
 //
 // Some changes about burst(b):
-//   - If b == 0, that means the limiter is unlimited capacity. default use in resource controller (burst within a capacity).
+//   - If b == 0, that means the limiter is unlimited capacity. default use in resource controller (burst with a rate within a unlimited capacity).
```
Comment (JmPotato) [inline client/resource_manager/client/limiter.go:65]:
```suggestion
//   - If b == 0, that means the limiter is unlimited capacity. default use in resource controller (burst with a rate within an unlimited capacity).
```

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

File: client/resource_manager/client/client.go [hunk 4]
```diff
@@ -133,15 +134,30 @@ func (c *ResourceGroupsController) Stop() error {
...
+		gc := tmp.(*groupCostController)
+		return gc, nil
+	}
 	group, err := c.provider.GetResourceGroup(ctx, name)
 	if err != nil {
 		return nil, err
 	}
-	log.Info("create resource group cost controller", zap.String("name", group.GetName()))
-	gc := newGroupCostController(group, c.config, c.lowTokenNotifyChan, c.groupNotificationCh)
+	if tmp, ok := c.groupsController.Load(name); ok {
```
Comment (nolouch) [inline client/resource_manager/client/client.go:146]: why twice checker in here? maybe add a comment here.

**Review by nolouch — APPROVED** [review]
rest LGTM

**Review by JmPotato — APPROVED** [review]

**Pull request merged**